### ISSUE TITLE
HybridForwardPass GPU embed lookup writes wrong values (post-#2)

### ISSUE THREAD

**Issue by pekkah** [pekkah/SharpInference#19]
## Status after 1369ca8 (2026-05-18)

Narrowed further to the GPU embed lookup. Phase-5 (final RmsNorm + output projection on GPU) is **ruled out** — verified by toggling `SHARPI_HYBRID_GPU_EMBED` and `SHARPI_HYBRID_GPU_OUTPUT` independently against the dense hybrid smoke test:

| Embed | Output | Result |
|---|---|---|
| CPU | CPU | PASS (workaround) |
| **GPU** | CPU | **FAIL** |
| CPU | **GPU** | PASS |
| **GPU** | **GPU** | **FAIL** |

Also ruled out this round:
- CPU-layer interaction path (fails with `CpuLayers=0` too)
- Buffer allocation order (deferring `_pinnedHidden` allocation didn't help)
- Stale `_gpuHidden` state (Clear + barrier before embed didn't help)

Workaround remains: `ShouldKeepFixedWeightsOnCpu` returns `true`. See [issue comment](https://github.com/pekkah/SharpInference/issues/19#issuecomment-4476739525) for the full repro.

---

Sub-issue of #3 with the suspect set narrowed after #2's descriptor-set fix landed.

## Status after 67a1e68

Commit 67a1e68 fixed the Vulkan §14.2.4 descriptor-set reuse hazard in `ComputePipeline.RecordWith` (per-dispatch DS allocation, recycled via `vkResetDescriptorPool` on each `BeginRecord` epoch). That fix resolved #2's MoE+hybrid NaN cascade end-to-end, so the DS hazard was real — but it is NOT the root cause of #3:

```bash
# All-GPU: works
dotnet run --project src/SharpInference.Cli -c Release -- \
  -m models/SmolLM2-1.7B-Instruct-Q4_K_M.gguf -g -1 --backend vulkan \
  -p "The capital of France is" --temp 0
# → "The capital of France is Paris." @ 162.6 t/s

# CPU-only: works
dotnet run --project src/SharpInference.Cli -c Release -- \
  -m models/SmolLM2-1.7B-Instruct-Q4_K_M.gguf -g 0 \
  -p "The capital of France is" --temp 0
# → "The capital of France is Paris." @ 48.5 t/s

# Hybrid with ShouldKeepFixedWeightsOnCpu=false (workaround OFF): broken
dotnet run --project src/SharpInference.Cli -c Release -- \
  -m models/SmolLM2-1.7B-Instruct-Q4_K_M.gguf -g 1 \
  -p "The capital of France is" --temp 0
# → Prefill: 36 tokens, 17.9 t/s | Decode: 0 tokens, 0.0 t/s
#   (model emits EOS immediately every Forward; logits are likely zero or near-zero
#    but technically finite, which is why the IsFinite-only test never caught this)
```

## Suspects (narrowed to two)

`ShouldKeepFixedWeightsOnCpu = false` only opens up two new code paths in `HybridForwardPass.Forward`:

1. **GPU `EmbedLookupQ4K` from Hybrid context** — the original suspect from #3. The same shader works correctly from `GpuForwardPass.Forward` for the same model + token, so it's a state-dependent bug, not a shader bug. Per the original #3 evidence, instrumentation showed 7 NaN and 815 out-of-magnitude values in the post-embed hidden state buffer for a 2048-element embedding row.

2. **Phase-5 GPU norm + output projection** — after the CPU layers complete, `HybridForwardPass.Forward` lines ~565-573 do `RmsNorm(_gpuHidden, _gpuHidden, _gpuOutputNorm)` then `GpuMatMul(_gpuLogits, _gpuOutputWeight, _gpuHidden)` then `RecordDownloadToStaging`. This path is also only exercised when the workaround is off. If the GPU output projection writes wrong logits, you'd get the same "EOS picked" failure mode.

Easiest way to differentiate: temporarily set `_gpuOutputWeight = null` (keeping `_gpuEmbedding` non-null) and verify whether the failure persists. If the failure goes away, the bug is in Phase-5, not in the embed lookup. If it persists, the embed lookup is at fault.

## What's been ruled out (from the original #3 investigation)

- Transfer/compute barrier issue (RecordComputeCopy is used throughout)
- Buffer aliasing (`_buffers` dict entries verified unique)
- Upload ordering
- Specific to the Q4_K shader (the F32 `EmbedLookup` shader produced identical garbage)
- Descriptor-set reuse (verified with this commit's DS fix in place)

## Test safety net

`tests/SharpInference.Tests.ForwardPass/VulkanShaderTests.cs::AssertHybridForwardPassProducesCoherentDecode` now checks all three: logits finite, argmax != EOS at first decode, and a 4-token greedy decode produces non-EOS variety. A future attempt at this fix will be caught by the existing tests if it produces "all EOS" failures.

## Workaround impact

`ShouldKeepFixedWeightsOnCpu` returning true forces CPU dequantization of one embedding row per token and runs the final norm + output projection on CPU. Cost: a few µs per token vs. millisecond-scale layer compute — effectively negligible. Hybrid decode rates from README's "What works today" matrix were measured with this workaround in place. Worth fixing for cleanliness but no user-visible perf is on the table.

**Comment by pekkah** [comment]
## Progress 2026-05-18: bug narrowed to embed lookup

Added independent toggles for the two suspect paths via env vars in `HybridForwardPass`:

- `SHARPI_HYBRID_GPU_EMBED=1` — force embed lookup on GPU
- `SHARPI_HYBRID_GPU_OUTPUT=1` — force final norm + output projection on GPU

Ran `HybridForwardPass_DenseSmallVocab_ProducesCoherentDecode` (SmolLM2, GpuLayers=1, CpuLayers=23) in all four combinations:

| Mode | Embed | Output | Result |
|---|---|---|---|
| A (default) | CPU | CPU | **PASS** |
| B | **GPU** | CPU | **FAIL** (49152 non-finite logits) |
| C | CPU | **GPU** | **PASS** |
| D | **GPU** | **GPU** | **FAIL** (49152 non-finite logits) |

**Conclusion:** the bug is entirely in the GPU embed lookup. Suspect #2 from this issue (Phase-5 norm+output projection) is ruled out — mode C passes.

## Additional ruled-out hypotheses (this session)

- **CPU-layer interaction path:** also fails with `GpuLayers=hp.NumLayers, CpuLayers=0` + `SHARPI_HYBRID_GPU_EMBED=1`. Bug is independent of the CPU↔GPU transition.
- **Buffer allocation order:** deferring `AllocatePinned(_pinnedHidden)` until after all weight uploads did not fix the failure.
- **Stale `_gpuHidden` state:** adding `Clear(_gpuHidden) + barrier` immediately before the embed dispatch did not fix the failure.

## Remaining suspects

The embed-lookup shader and its bindings are identical between `HybridForwardPass` and `GpuForwardPass`. The descriptor-set hazard from #2 is also already fixed (per-dispatch DS allocation in `ComputePipeline.RecordWith`). What's left:

- Something about the pipeline-cache or shader-module lifecycle differing between class init paths
- Or a Vulkan validation-layer-silent error during the first dispatch in HybridForwardPass

Next step worth trying: enable `VK_LAYER_KHRONOS_validation` with verbose logging and re-run mode D. If the layer is silent, instrument the embed dispatch with a download-and-print right after it and compare byte-for-byte with the same dispatch from `GpuForwardPass`.

Workaround remains in place; cost is negligible.
